tools/ja-rule/ja-rule-controller.cpp

```diff
@@ -0,0 +1,440 @@
+             << upper;
+    m_device->SendMessage(OpenLightingDevice::RDM_DUB, data, rdm_length);
+
+    // TODO(simon): do we need to check there isn't already one in-flight?
+    m_branch_callback = branch_complete;
+  }
+
+  void ResetDevice() {
+    m_device->SendMessage(OpenLightingDevice::RESET_DEVICE, NULL, 0);
+  }
+
+  void NewMessage(const Message& message) {
+    OLA_INFO << "Got message with command "
+             << static_cast<int>(message.command);
+
+    switch (message.command) {
+      case OpenLightingDevice::RDM_DUB:
+        HandleDUBResponse(message);
+        break;
+      case OpenLightingDevice::RDM_REQUEST:
+        HandleRDM(message);
+        break;
+      default:
+        OLA_WARN << "Unknown command: " << ToHex(message.command);
+    }
+
+    if (message.flags & LOGS_PENDING_FLAG) {
+      OLA_INFO << "Logs pending!";
+    }
+    if (message.flags & FLAGS_CHANGED_FLAG) {
+      OLA_INFO << "Flags changed!";
+    }
+    if (message.flags & MSG_TRUNCATED_FLAG) {
+      OLA_INFO << "Message truncated";
+    }
+  }
+
+ private:
+  OpenLightingDevice *m_device;
+  const UID m_our_uid;
+  ola::SequenceNumber<uint8_t> m_transaction_number;
+  MuteDeviceCallback *m_mute_callback;
+  UnMuteDeviceCallback *m_unmute_callback;
+  BranchCallback *m_branch_callback;
+
+  bool CheckForDevice() const {
+    if (!m_device) {
+      OLA_INFO << "Device not present";
+    }
+    return m_device != NULL;
+  }
+
+  void HandleDUBResponse(const Message& message) {
+    if (m_branch_callback) {
+      const uint8_t *data = NULL;
+      unsigned int size = 0;
+      if (message.payload && message.payload_size > 1) {
+        data = message.payload + 1;
+        size = message.payload_size - 1;
+      }
+      BranchCallback *callback = m_branch_callback;
+      m_branch_callback = NULL;
+      callback->Run(data, size);
+    }
+  }
+
+  void HandleRDM(const Message& message) {
+    if (m_unmute_callback) {
+      UnMuteDeviceCallback *callback = m_unmute_callback;
+      m_unmute_callback = NULL;
+      callback->Run();
+      return;
+    }
+
+    if (m_mute_callback) {
+      // TODO(simon): inflate the actual RDM response here. Right now we treat
+      // any response as good.
+      bool ok = message.payload_size > 1;
+      MuteDeviceCallback *callback = m_mute_callback;
+      m_mute_callback = NULL;
+      callback->Run(ok);
+    }
+  }
+
+  DISALLOW_COPY_AND_ASSIGN(ControllerImpl);
+};
+
+class Controller : public ola::rdm::DiscoverableRDMControllerInterface {
+ public:
+  Controller(OpenLightingDevice *device, const UID &controller_uid)
+    : m_controller_impl(device, controller_uid),
+      m_discovery_agent(&m_controller_impl) {
+  }
+
+  ~Controller() {
+  }
+
+  void UpdateDevice(OpenLightingDevice* device) {
+    m_controller_impl.UpdateDevice(device);
+  }
+
+  void SendRDMRequest(const RDMRequest *request,
+                      ola::rdm::RDMCallback *on_complete) {
+    // TODO(simon): implement me!
```

Comment:
Is this still your WIP? Otherwise it should log a warning or error message if it hits this.